### PR TITLE
implement bidirectional transcoding to replace interpreters

### PR DESCRIPTION
**Pull request by pezra**
See readme and module docs for usage.

### CODE REVIEW

File: lib/exhal/document.ex [hunk 1]
```diff
@@ -8,32 +8,56 @@ defmodule ExHal.Document do
 
   defstruct properties: %{}, links: %{}, client:
 
+
   @doc """
     Returns a new `%ExHal.Document` representing the HAL document provided.
     """
-  def parse(client, hal_str) do
+  def parse(hal_str, client \\ ExHal.client)
```
Comment (alakra) [inline lib/exhal/document.ex]: I haven't used this form, what does it mean?

**Comment by alakra** [comment]
:+1:

File: lib/exhal/document.ex (hunk 1 above)
Comment (pezra) [inline lib/exhal/document.ex]: it is called a "function head" and it's causes a compiler warning when there are multiple function clauses and only some of them have default values.

Comment (alakra) [inline lib/exhal/document.ex]: Ah ok, I'd seen this when defining protocols, but not elsewhere. Cool!

**Comment by glassbead0** [comment]
I don't understand macros enough to provide useful feedback, but I like the new functionality.

**Pull request merged**